src/time_integrator.f90

```diff
@@ -5,73 +5,158 @@ module m_time_integrator
     else
-      constructor%nvars = 3
+      init%nvars = 3
     end if
 
-    constructor%nolds = 0
+    init%istep = 1
+    init%nolds = init%order - 1
 
-    allocate (constructor%olds(constructor%nvars, constructor%nolds))
+    ! allocate memory
+    allocate (init%olds(init%nvars, init%nolds))
+    allocate (init%curr(init%nvars))
+    allocate (init%deriv(init%nvars))
 
     ! Request all the storage for old timesteps
-    do i = 1, constructor%nvars
-      do j = 1, constructor%nolds
-        constructor%olds(i, j)%ptr => allocator%get_block(DIR_X)
+    do i = 1, init%nvars
+      do j = 1, init%nolds
+        init%olds(i, j)%ptr => allocator%get_block(DIR_X)
       end do
     end do
 
-  end function constructor
+  end function init
 
   subroutine step(self, u, v, w, du, dv, dw, dt)
     implicit none
 
-    class(time_intg_t), intent(in) :: self
-    class(field_t), intent(inout) :: u, v, w
-    class(field_t), intent(in) :: du, dv, dw
+    class(time_intg_t), intent(inout) :: self
+    class(field_t), target, intent(inout) :: u, v, w
+    class(field_t), target, intent(in) :: du, dv, dw
 
     real(dp), intent(in) :: dt
 
-    call self%backend%vecadd(dt, du, 1._dp, u)
-    call self%backend%vecadd(dt, dv, 1._dp, v)
-    call self%backend%vecadd(dt, dw, 1._dp, w)
+    ! assign pointer to variables
+    self%curr(1)%ptr => u
+    self%curr(2)%ptr => v
+    self%curr(3)%ptr => w
 
+    ! assign pointer to variables
+    self%deriv(1)%ptr => du
+    self%deriv(2)%ptr => dv
+    self%deriv(3)%ptr => dw
+
+    call self%adams_bashforth(dt)
+
+    ! increment step counter
+    self%istep = self%istep + 1
   end subroutine step
 
-  subroutine adams_bashford_1st(vels, olds, coeffs)
-    type(flist_t) :: vels(:), olds(:)
-    real :: coeffs(:)
+  subroutine adams_bashforth(self, dt)
+    class(time_intg_t), intent(inout) :: self
+    real(dp), intent(in) :: dt
 
-    !call vec_add(vels, olds, coeffs)
-  end subroutine adams_bashford_1st
+    integer :: i, j
+    integer :: order
+
+    order = min(self%istep, self%order)
+    do i = 1, self%nvars
+      ! update solution
+      call self%backend%vecadd(self%coeffs(1, order)*dt, &
+                               self%deriv(i)%ptr, &
+                               1._dp, self%curr(i)%ptr)
+      do j = 2, order
+        call self%backend%vecadd(self%coeffs(j, order)*dt, &
+                                 self%olds(i, j - 1)%ptr, &
+                                 1._dp, self%curr(i)%ptr)
+      end do
+
+      ! rotate pointers
+      if (order < self%order) then
+        ! for startup
+        if (self%istep > 1) then
+          call rotate(self%olds(i, :), order)
+        end if
+      else
+        ! after startup
+        if (self%order > 2) then
+          call rotate(self%olds(i, :), order - 1)
+        end if
+      end if
+
+      ! update olds(1) with new derivative
+      if (self%order > 1) then
+        call self%backend%vecadd(1.0_dp, self%deriv(i)%ptr, 0._dp, &
```

Comment:
We can potentially do a pointer switch here too I think. Not very important but saves a few `vecadd`s. If you decide to give this a go please run AB3 again with the TGV to make sure it doesn't break anything, and also check out the memory usage too. A mistake in pointer switch might lead to constant allocations at each time step.

Reply:
I think I will do that in the next PR about Runge-Kutta. I am now thinking to add a swap routine.

Reply:
One difference is the `self%deriv` pointers point to the input parameters `du, dv, dw`. And in `solver.f90` there is this function call to
`self%backend%allocator%release_block(du)`

Is this going to cause problem?

With `self%olds`, everything is internal to the time integrator type.

Reply:
Its completely fine provided that `du` points to a field. So if you do
```
ptr => olds(?)
olds(?) => du
du => ptr
```
I think it should work. Probably requires `du` to be `intent(inout)`.

Reply:
Ah, you work on them via the `flist_t` `deriv`, so it can be more tricky than the code snipped above.

Reply:
Yeah, `flist_t olds` and  `flist_t deriv`  are going to be swaped, but `du` still point to the same memory location, which will be released just after...